### PR TITLE
Render improvement

### PR DESCRIPTION
**Pull request by taooceros**
Change UI rendering logic to fix image flickering (the major fix is by removing the isAsync for image binding)
This comes from @bao-qian 's change in Wox, [#2899](https://github.com/Wox-launcher/Wox/pull/2899). Thank you for introducing the fix for image flickering.

The other change is majoring makes the code simpler and more readable, and remove duplicate logic.

Also fix that sometimes the selected index is not first when updating result views. https://social.msdn.microsoft.com/Forums/vstudio/en-US/5ff71969-f183-4744-909d-50f7cd414954/binding-a-tabcontrols-selectedindex-not-working?forum=wpf

### CODE REVIEW

**Comment by jjw24** [comment]
Hmmm, getting this odd behaviour...
image

**Comment by jjw24** [comment]
Closes https://github.com/Flow-Launcher/Flow.Launcher/issues/42

**Comment by taooceros** [comment]
Here should work better.

**Comment by jjw24** [comment]
When I am query, i am missing results. It seems it only gets the top portion of the results list, evident when i use Explorer plugin to browse folders.

**Comment by taooceros** [comment]
> When I am query, i am missing results. It seems it only gets the top portion of the results list, evident when i use Explorer plugin to browse folders.

That's because I use the MaxResult to limit the number of results to be added into result list. If that's not intended, I can remove it.
image

**Comment by jjw24** [comment]
> > When I am query, i am missing results. It seems it only gets the top portion of the results list, evident when i use Explorer plugin to browse folders.
> 
> That's because I use the MaxResult to limit the number of results to be added into result list. If that's not intended, I can remove it.
> image
> 
> 

Oh I see. Let's keep the original behaviour, let it flow :)

**Comment by taooceros** [comment]
> Oh I see. Let's keep the original behaviour, let it flow :)

ok!

**Comment by jjw24** [comment]
so far I came across intermittent issues:
- in the result list, sometimes highlighting does not show, so I couldnt just type the result and then press enter on the 1st result
- the highlighting sometimes jumped to 2nd or 3rd result

**Comment by taooceros** [comment]
> in the result list, sometimes highlighting does not show, so I couldnt just type the query and then press enter on the 1st result

This may because of the delay I added for batch processing, let's make it shorter and take a try.

> the highlighting sometimes jumped to 2nd or 3rd result

This is quite weird....and I haven't met that situation. Let's take a look after I reduce the delay time.

**Comment by taooceros** [comment]
UPDATE
> the highlighting sometimes jumped to 2nd or 3rd result

That's because some of the plugin query is slower, so that their result view is updated later than others, and the selection has not been updated, so that the selection is not on them even though their score is higher. I will try to update selection every time result view is change.

**Comment by jjw24** [comment]
just used the latest changes, hmm still getting this incorrect result highlighting occasionally:
image

**Comment by taooceros** [comment]
Weird......I haven't met that situation after the commit. Let me take a look again later,

**Comment by jjw24** [comment]
also i am experiencing random delays occasionally.. strange

**Comment by taooceros** [comment]
I have met that it will stunt at few second at the first query, and then the query is fine. Well....I don't understand why that is happen.

**Comment by taooceros** [comment]
I added a null check when processing current result view, which seems solve some stunt for logging.
At the same time of testing, I found that the explorer plugin seems sometimes fail at query some of the query with special character.
image

**Comment by jjw24** [comment]
> I added a null check when processing current result view, which seems solve some stunt for logging.

Cool, I will update to the latest and test

> At the same time of testing, I found that the explorer plugin seems sometimes fail at query some of the query with special character.
> image
> 

Can you put this into a quick issue please so I can remmeber to get around to fixing it.

**Comment by taooceros** [comment]
> Can you put this into a quick issue please so I can remmeber to get around to fixing it.

Sure

**Comment by jjw24** [comment]
> I added a null check when processing current result view, which seems solve some stunt for logging.
> At the same time of testing, I found that the explorer plugin seems sometimes fail at query some of the query with special character.
> image

sorry mate, it seems the stunt/delay is still there when i query, both query and result list freezes, quite reproducible now with queries that return longer list of results. use Explorer to test

**Comment by taooceros** [comment]
Thank you for pointing that out, let me try on large set of query result. That may be a good way for me to debug the performance of the change.

**Comment by taooceros** [comment]
> sorry mate, it seems the stunt/delay is still there when i query, both query and result list freezes, quite reproducible now with queries that return longer list of results. use Explorer to test

Weird, why doesn't I get the stunt when the resultview is large. Could you tell me is it the first-time query? I do encounter stunt when query at the first time.

image
You can see from the scroll bar that the result count is large, but I haven't encountered the stunt. It appears around 20ms, I think? (I think the explorer plugin has set a limit of 100 results so this is the largest result set I can get).

Let me check what can lead the listview stunt.

**Comment by taooceros** [comment]
I found out the reason of stunting for me, which is the image loading hasn't been done, and because the image binding `isAsync` is removed for fixing image flickering, the ui stunt at the first time query since the image hasn't been done.

**Comment by jjw24** [comment]
Awesome, good find

**Comment by jjw24** [comment]
> Weird, why doesn't I get the stunt when the resultview is large. Could you tell me is it the first-time query? I do encounter stunt when query at the first time.

I'm getting it after doing the same query several times... But not always consistently

**Comment by taooceros** [comment]
Sorry, the last commit isn't behaving like what it has described. Please Ignore it.
> I'm getting it after doing the same query several times... But not always consistently

Let me try if I can add the default image first, and then change the image to the image that it should like.

**Comment by taooceros** [comment]
Let's try this one!

**Comment by taooceros** [comment]
Weird....why do the build fail again? Is anything in the building environment change?

**Comment by jjw24** [comment]
So just tested latest, i get the stunt/delay + issues, all happen occasionally:
1. on start up after one character typed
2. still getting the same stunt/delay after typing a certain query, for me it's using Explorer with shortcut + "docker"
3. and on rare occasions result list doesnt seem to be updating correctly:
image

**Comment by taooceros** [comment]
The last commit is quite magic for me toward multitasking. It seems that I originally use global token assign to each result update, which results unintended token assignment so that the result update doesn't behave properly.

This fixs my issue of unintended update.

**Comment by taooceros** [comment]
The last commit seems reduce my startup stunts. But require more testing whether it is effective.
The reason for adding that is WPF hasn't supported multi-item operation in listview binding, but simply notice reset at startup seems cause stunt issue? I am not quite sure about that.
https://github.com/dotnet/runtime/issues/18087

**Comment by jjw24** [comment]
> So just tested latest, i get the stunt/delay + issues, all happen occasionally:
> 
> 1. on start up after one character typed
> 2. still getting the same stunt/delay after typing a certain query, for me it's using Explorer with shortcut + "docker"
> 3. and on rare occasions result list doesnt seem to be updating correctly, (e is my shortcut to trigger Explorer):

Using the latest commit 8f6f2fb, no longer getting the above problems.

However result seems to jump around, especially evident when using WebSearch, actionkeyword set to * and with suggestions enabled. The original behaviour is the suggestions result is updated while the search stays atop. The WebSearch suggestions result has a bit of delay so in the code it is set to update results list view once the api call is done, maybe (have not looked at the code yet) this bit of behaviour is not honored?

**Comment by taooceros** [comment]
> However result seems to jump around occasionally, evident when using WebSearch, actionkeyword set to * and with suggestions enabled. The original behaviour is the suggestions result is updated while the search stays atop.

hmmm not sure what you mean? Are you mean the selected result jump around? I haven't got issue when only assigning one search source to * as action keyword. However, when multiple search source are bounded together, it seems that it works unintended.

> The original behaviour is the suggestions result is updated while the search stays atop.

That's true, I have checked the code, and it has used the event driven callback to update the later result.

**Comment by taooceros** [comment]
Multiple Search Source may due to the design issue of WebSearch, I will check that out later.

**Comment by taooceros** [comment]
UPDATE: It seems that may due to the unintended assignment of global token, which is assigned later, so that it is not cancelled, which leads to the results flow. Let me check how to solve that issue.

**Comment by taooceros** [comment]
@jjw24 fix that.

**Comment by jjw24** [comment]
Amazing, will take it for a spin now.

Also i havent looked at the code, and just as a suggestion only, is it possible to pull out the fix for image flicking issue as a standalone pr while this result list improvement is worked on? or are they tied up together

**Comment by taooceros** [comment]
> Amazing, will take it for a spin now.
> 
> Also i havent looked at the code, and just as a suggestion only(totally up to you), is it possible to pull out the fix for image flicking issue as a standalone pr while this result list improvement is worked on? or are they tied up together

Ok, that's possible, I believe that the flickering issue is only caused by the IsAsync property in xaml image tag, so the fix toward that only include removing async loading and adding customized LazyAsync class for loading default image before the image is loaded in cache.

**Comment by jjw24** [comment]
> > However result seems to jump around occasionally, evident when using WebSearch, actionkeyword set to * and with suggestions enabled. The original behaviour is the suggestions result is updated while the search stays atop.

this is sort of what i mean, can now reliably reproduce with 'linux academ'. it should be the first result in the list rather than 'linux academy'

image

**Comment by taooceros** [comment]
> > > However result seems to jump around occasionally, evident when using WebSearch, actionkeyword set to * and with suggestions enabled. The original behaviour is the suggestions result is updated while the search stays atop.
> 
> this is what i mean, can now reliably reproduce with 'linux academ'. it should be the first result in the list rather than 'linux academy'
> 
> image
> 
> repro:
> websearch => *, enable suggestions
> 
> type 'linux academ', first time it displays first result 'linux academ' and highlights the first result (which is correct). Select second result 'linux academy', then retype 'linux academ' and you get the screenshot above.
> 
> Correct behaviour should be: highlighting on first result 'linux academy', as is consistent with dev branch behaviour.

Oh, I misunderstand your problem. I will check that today.

**Comment by jjw24** [comment]
just merged dev back into this PR. Please check if the merge changes are ok

**Comment by taooceros** [comment]
> just merged dev back into this PR. Please check if the merge changes are ok

Seems great.

**Comment by taooceros** [comment]
@jjw24 fix it by also sync selected item with selected index when adding results.

**Comment by jjw24** [comment]
Let me give it a spin

**Comment by taooceros** [comment]
Found out that the context menu cannot be displayed.....not sure what cause that. Convert to draft until I fix it.

**Comment by taooceros** [comment]
Found out that the context menu cannot be displayed.....not sure what cause that. Convert to draft until I fix it.

**Comment by jjw24** [comment]
no problem.

I had fair bit of issues with the commits a640ead. i found 6b96228 to be quite stable with only one issue with Explorer plugin where if it doesnt find a result, flow will then not display any plugin's results, no error in logs either.

**Comment by taooceros** [comment]
> no problem.
> 
> I had fair bit of issues with the commit a640ead. i found 6b96228 to be quite stable with only one issue with Explorer plugin where if it doesnt find a result, flow will then not display any plugin's results, no error in logs either.

Yes you are right, it seems that observable collection has done something special to make sure binding works better. I do not fully understand why, but I will move it back, and it seems much more stable.
I have met an issue that I want to ask whether you have met when trying this branch. Sometime after quick typing, the resultlist won't be able to display, and the whole data thread block, which seems due to invoke collection change event. Not sure whether it is possible to false reset when invoking stunt.

**Comment by taooceros** [comment]
Change back to ObservableCollection works well.

**Comment by jjw24** [comment]
> I have met an issue that I want to ask whether you have met when trying this branch. Sometime after quick typing, the resultlist won't be able to display, and the whole data thread block, which seems due to invoke collection change event. Not sure whether that only appear when I debug or it is an issue. 

Yep I had that with commit 6b96228  I was able to consistently replicate with explorer plugin for some reason when it doesn't find a result

**Comment by taooceros** [comment]
> Yep I had that with commit 6b96228 I was able to consistently replicate with explorer plugin for some reason when it doesn't find a result

Are you compiling it as debug? Is any error throw or simply no results show? Let me investigate it more. Would you please tell me is it anyway to replicate it consistently?

> Does this solve the data thread block?

I wrap the data update event inside a Task.Run, and wait it only for at most 300ms. This should solve the block issue. It seems that I haven't encounter the same issue after changing that. But what's weird is that the timeout exception also hasn't been thrown in those testing.

**Comment by taooceros** [comment]
Take Back my word. The thread block isn't caused by updating the observable collection, but due to the popped-out exception for checking whether unacceptable cancellation token appear in the update queue. This cause the Task stop but not throwing exception since it is wrap in the task.run. I change that to only throw it on debug and recreate the task if any exception is throw.

**Comment by jjw24** [comment]
>Are you compiling it as debug? Is any error throw or simply no results show? Let me investigate it more. Would you please tell me is it anyway to replicate it consistently?

i test with a release build actually. 

Should i test again now after the updates?

**Comment by taooceros** [comment]
> >Are you compiling it as debug? Is any error throw or simply no results show? Let me investigate it more. Would you please tell me is it anyway to replicate it consistently?
> 
> i test with a locally compiled release build actually. 
> 
> Should i test again now after the updates?

Well, I don't think the commit above seems able to fix the empty result when explorer plugin cannot find result. Howeve, you can take a try on debugging it and seems what's happening when explorer plugin cannot find result since I haven't successfully reproduced that. The error handling has been added so that debugging may be useful to find out what happens. I will take a try on using release version and see what happens if error throw, since debug version will simply throw the error and not able to see what happen after error.

**Comment by taooceros** [comment]
Well....I think this possibly caused by the long query time of explorer plugin in some special case, and it is not able to cancel itself since no CancellationToken is passed to it, which is causing thread deadlock in thread pool……
image
You can see from the picture that there is a large number of works that is queued to wait to be done, and most of them are in Explorer plugin.

**Comment by jjw24** [comment]
hmm ok, seems different. Repro my one:

1. Was using  commit 6b96228.
2. Using locally compiled release build
3. Assign an action key to explorer, mine was 'e'
4. Trigger action key word, and type random letters till explorer no longer show result
5. Delete entire query and retype anything and no results show up

Side note using the above steps if Explorer is able to get a result, then everything is still ok

**Comment by taooceros** [comment]
> hmm ok, seems different. Repro my one:
> 
> 1. Was using  commit [6b96228](https://github.com/Flow-Launcher/Flow.Launcher/commit/6b9622844c3beda0a32c564f906787f2a30d0df4).
> 2. Using locally compiled release build
> 3. Assign an action key to explorer, mine was 'e'
> 4. Trigger action key word, and type random letters till explorer no longer show result
> 5. Delete entire query and retype anything and no results show up
> 
> Side note using the above steps if Explorer is able to get a result, then everything is still ok

I believe we are encountering the same issue. When encounter complex query, the explorer takes too long to query, which is unable to stop when next query comes, and finally cause too much threads in threadpool to wait previous search, so that other task won't be able to run even though queued.

Is it possible to add cancellation to explorer plugin, and what do you think about #233 ?

**Comment by taooceros** [comment]
Sorry, incidentally close the pr.

**Comment by jjw24** [comment]
@taooceros is this good for another round of testing?

**Comment by taooceros** [comment]
> @taooceros is this good for another round of testing?

Let me check whether merging dev create some issue.

**Comment by taooceros** [comment]
ready for another round of testing now

**Comment by jjw24** [comment]
hmmmmm

image

image

**Comment by taooceros** [comment]
Hmmm that's weird.... Haven't encountered before, will take a look now.

**Comment by taooceros** [comment]
fix, this is due to merge issue




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Flow-Launcher/Flow.Launcher"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2021年1月10日(星期天) 上午10:47
收件人:&nbsp;"Flow-Launcher/Flow.Launcher"<Flow.Launcher@noreply.github.com&gt;;
抄送:&nbsp;"天煞"<1569418337@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [Flow-Launcher/Flow.Launcher] Render improvement (#195)





 
hmmmmm

**Comment by jjw24** [comment]
dups appearing for web search:
image

**Comment by taooceros** [comment]
> dups appearing for web search:
> image

hmmm this is really hard to reproduce....... I have only encountered once unintentionally.

**Comment by taooceros** [comment]
Are they appearing at the same time or the second one appearing later?

**Comment by jjw24** [comment]
they are appearing at the same time when using websearch plugin

**Comment by jjw24** [comment]
i am running it directly from the branch in visual studio

**Comment by jjw24** [comment]
remove output, remove roaming data location, then run from vs, search 'g a' and i get it instantly

**Comment by taooceros** [comment]
hmmm still cannot reproduce.... Are you using Google as search suggestion or none? I have tried Baidu and Bing, and both are fine.

**Comment by jjw24** [comment]
None. Basically all default settings

**Comment by taooceros** [comment]
> None. Basically all default settings

Fix it. It is due to result update event create duplicate result.

**Comment by jjw24** [comment]
@taooceros could you merge dev in and resolve conflict please

**Comment by taooceros** [comment]
> @taooceros could you merge dev in and resolve conflict please

Sure

File: Flow.Launcher/ViewModel/ResultViewModel.cs [hunk 1]
```diff
@@ -131,6 +128,7 @@ public override bool Equals(object obj)
             }
         }
 
+
```
Comment (jjw24) [inline Flow.Launcher/ViewModel/ResultViewModel.cs:131]: lol how come ur prs always have random additional white space

**Review by jjw24 — APPROVED** [review]

**Pull request merged**